### PR TITLE
GREP: avoids tedit-file formatting, font change chars in Lisp source files, adds TGREP

### PR DESCRIPTION
**Pull request by rmkaplan**
The tedit-file part works only in 4th round or later (see discussion in #1496).

New entry TGREP puts the results in a scrollable TEDIT stream.  

This also suppresses characters below (CHARCODE LF) in a Lisp source file.

A cosmetic different: filenames are preceded by a blank line and printed in bold.

### CODE REVIEW

File: lispusers/GREP [hunk 1]
```diff
@@ -1,32 +1,34 @@
...
-")
 
 (PRETTYCOMPRINT GREPCOMS)
 
-(RPAQQ GREPCOMS ((FNS DOGREP GREP PHONE)
-                 (INITVARS (PHONELISTFILES))))
+(RPAQQ GREPCOMS
+       [(FNS DOGREP GREP LISPSOURCEOUTCHARFN TGREP)
+        (P (MOVD? 'NILL 'TEDIT.FORMATTEDFILEP))
+        (DECLARE%: EVAL@COMPILE DONTCOPY (FILES (FROM VALUEOF (MEDLEYDIR "loadups"))
```
Comment (masinter) [inline lispusers/GREP:17]: could we load exports.all more consistently, or move the functionality into tedit or ...

Comment (nbriggs) [reply lispusers/GREP:17]: Yeah, not a big fan of `(FROM VALUEOF (MEDLEYDIR "loadups"))` embedded in the COMS.

**Comment by rmkaplan** [comment]
On that front, I tend to say (FILES (FROM loadups)…), because MEDLEYDIR is used basically to set up DIRECTORIES, LISPUSERSDIRECTORIES…  Is that better?

I think the only need for this is to get the stream declaration, so it can change the OUTCHARFN to suppress font change characters.  Wouldn’t be necessary if STREAMPROP saw the external format fields as properties, so that the format behavior can be changed without creating a whole new format.

**Comment by masinter** [comment]
the font control-codes are just treated as a SEPRCHAR ... is this going to be used for EDITCALLERS too?
(This was originally the 'phone' LispUsers package, wasn't it?)

**Comment by rmkaplan** [comment]
I got rid of the dependence on the stream record, by just suppressing lower-control codes in all cases. It could be made sensitive to LISPSOURCEFILEP, but that's probably not too valuable.  (Note that the prior set up had a terrible bug, anyway, this supercedes).

**Comment by nbriggs** [comment]
Is it expected that if you do, when connected to `medley/library`
```
(GREP "FOO" "BROWSER.TEDIT")
```
you get a `TYPE-MISMATCH` error for `NIL is not a NUMBER` ?
[grep-error.txt](https://github.com/Interlisp/medley/files/14624663/grep-error.txt)

**Comment by rmkaplan** [comment]
That’s interesting.

It’s trying to look just at the “logical” characters of the Tedit file, and it is passing in the byte position of the end of the plaintext part of the file to prevent the search from going uselessly into the formatting section (unless you were looking, I suppose, for the name of a font or the internals of an image object—then use the real grep).

But the branch for Tedit files should be outside of the search loop.  After the with-open-file opens it as a simple stream, if it satisfies the Tedit test, the stream should be coerced to a text stream and the search should run on that.  FILEPOS will probably break when it tries to map an image object through the upper-case array, so the stream should also set the property that returns a special character code for objects.

**Comment by rmkaplan** [comment]
Try it now

**Comment by nbriggs** [comment]
Works on BROWSER.TEDIT, but try it on lispusers/BLACKBOX.TEDIT -- I got a `HELP! FILE BUFFER FOR NON-FILE PIECE`

**Comment by nbriggs** [comment]
When it finds the text in a TEdit file, it reports it as being found in (for example)
`(from #<IO Text Stream/137,5600)`
rather than the original file name.

**Comment by nbriggs** [comment]
Test case:
```
CONN medley/lispusers
(GREP "FOO" (DIRECTORY "*.TEDIT"))
```

**Comment by rmkaplan** [comment]
This is a problem in Tedit, Grep is OK.

(It seems to get in the wrong state when binning from a file is followed by peeking at an object.)

**Comment by rmkaplan** [comment]
This is fixed in last commit.  (the HELP is an independent problem)

**Comment by nbriggs** [comment]
OK, looks good then.  I discovered by accident that you have to be careful not to give it PATHNAMEs instead of atoms or strings for the files - grepping in `(CL:DIRECTORY "*")` fails, but that's `DIRECTORYNAMEP` failing because `PACKFILENAME.STRING` is unhappy.
image

**Review by nbriggs — APPROVED** [review]
Looks better, see comments in conversation.

**Comment by rmkaplan** [comment]
Apart from fixing the lowe-level peekbin bug in Tedit, Grep could probably do more with image objects.

The line detection just ignores them for the match (you presumably can’t match on an object, and FILEPOS doesn’t like them), but it would be relatively easy to produce some representation of a particular image object on a line that elsewhere has a matching substring—a GETFILEPTR and a TEDIT.NTHCHARCODE.  We could at least show its getfn, maybe in italics, or even run the  display function of the image object so the whole thing appears in the output stream (if output is going to an image stream).  Or just run its PREPRINTFN.

**Pull request merged**